### DISCUSSION TITLE
TypeError: 'int' object is not iterable

### DISCUSSION THREAD

**Discussion by deven367** [explosion/spaCy#7524] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

I am trying to train my own NER model with custom labels. My training data looks something like this:
```
 [('&lt;', {'entities': [0, 3, 'O']}),
 ('watching', {'entities': [0, 7, 'O']}),
 ('the', {'entities': [0, 2, 'O']}),
 ('VMA', {'entities': [0, 2, 'B-creative-work']}),
 ('pre-show', {'entities': [0, 7, 'O']}),
 ('again', {'entities': [0, 4, 'O']}),
 ('lol', {'entities': [0, 2, 'O']})]
```
This is the code I am trying to use to train my data.
```
other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
with nlp.disable_pipes(*other_pipes):  # only train NER
    for itn in range(n_iter):
        random.shuffle(a)
        losses = {}
        batches = minibatch(a, size=compounding(4., 32., 1.001))
        for batch in batches:
            texts, annotations = zip(*batch)
            nlp.update(texts, annotations, sgd=optimizer, drop=0.35,
                        losses=losses)
        print('Losses', losses)
```
and I keep getting this error!
```
TypeError                                 Traceback (most recent call last)
<ipython-input-56-01ff2c4bf5ea> in <module>()
      9             texts, annotations = zip(*batch)
     10             nlp.update(texts, annotations, sgd=optimizer, drop=0.35,
---> 11                         losses=losses)
     12         print('Losses', losses)

1 frames
/usr/local/lib/python3.7/dist-packages/spacy/language.py in _format_docs_and_golds(self, docs, golds)
    470                     err = Errors.E151.format(unexp=unexpected, exp=expected_keys)
    471                     raise ValueError(err)
--> 472                 gold = GoldParse(doc, **gold)
    473             doc_objs.append(doc)
    474             gold_objs.append(gold)

gold.pyx in spacy.gold.GoldParse.__init__()

gold.pyx in spacy.gold.biluo_tags_from_offsets()

TypeError: 'int' object is not iterable
```

## Your Environment

* Operating System: Google colab
* Python Version Used: 3.7.10
* spaCy Version Used: 2.2.4
* Environment Information: just a notebook on Google Colab

**Comment by polm** [comment, marked as answer]
This is happening because you are passing a single entity where a list is expected, see [the old docs](https://v2.spacy.io/usage/training#training-data:~:text=create%20six%20training%20sentences%20with%20eight%20entities%20in%20total) for an example of correctly formatted data. It looks like this:

```
train_data = [
    ("Uber blew through $1 million a week", [(0, 4, 'ORG')]),
    ("Android Pay expands to Canada", [(0, 11, 'PRODUCT'), (23, 30, 'GPE')]),
    ("Spotify steps up Asia expansion", [(0, 8, "ORG"), (17, 21, "LOC")]),
    ("Google Maps launches location sharing", [(0, 11, "PRODUCT")]),
    ("Google rebrands its business apps", [(0, 6, "ORG")]),
    ("look what i found on google! 😂", [(21, 27, "PRODUCT")])]
```

That said, you have a more serious problem here - NER training data is supposed to be complete sentences, but you have single words. The model can't learn anything useful from training data like that. If all you have is a list of words look at the [rule-based matching docs](https://spacy.io/usage/rule-based-matching). 

**EDIT:** Actually, it looks like you had a complete sentence, but you've split it into words for annotations somehow? The example [here](https://v2.spacy.io/usage/training#example-train-ner:~:text=Who%20is%20Shaka%20Khan) in the old docs is close to what you have but note that each line is a sentence, not a single token.

Also, spaCy 2.2.4 is pretty old at this point. I'd recommend you upgrade to the recent v3 if possible, or, if you need to keep using v2, at least to the latest 2.3 release.

**Reply by deven367** [reply]
Oh! I did not know about the fact of giving full sentences to the NER model. I will try to use a different approach to solve this problem. This helped a lot. Thank you!
